src/main/java/com/gregtechceu/gtceu/common/machine/multiblock/electric/DistillationTowerMachine.java

```diff
@@ -0,0 +1,358 @@
+                Map<RecipeCapability<?>, List<Content>> out = Map.of(ItemRecipeCapability.CAP, items);
+                result = recipe.matchRecipeContents(IO.OUT, holder, out, false);
+                if (!result.isSuccess()) return result;
+            }
+
+            if (!applyFluidOutputs(recipe, FluidAction.SIMULATE)) {
+                return GTRecipe.ActionResult.fail(() -> Component.translatable("gtceu.recipe_logic.insufficient_out")
+                        .append(": ")
+                        .append(FluidRecipeCapability.CAP.getName()));
+            }
+
+            return GTRecipe.ActionResult.SUCCESS;
+        }
+
+        private void updateWorkingRecipe(GTRecipe recipe) {
+            if (recipe.recipeType == GTRecipeTypes.DISTILLERY_RECIPES) {
+                this.workingRecipe = recipe;
+                return;
+            }
+
+            this.workingRecipe = recipe.copy();
+            var contents = recipe.getOutputContents(FluidRecipeCapability.CAP);
+            var outputs = getMachine().getFluidOutputs();
+            List<Content> trimmed = new ArrayList<>(12);
+            for (int i = 0; i < Math.min(contents.size(), outputs.size()); ++i) {
+                if (!(outputs.get(i) instanceof VoidFluidHandler)) trimmed.add(contents.get(i));
+            }
+            this.workingRecipe.outputs.put(FluidRecipeCapability.CAP, trimmed);
+        }
+
+        @Override
+        protected boolean handleRecipeIO(GTRecipe recipe, IO io) {
+            if (io != IO.OUT) {
+                if (super.handleRecipeIO(recipe, io)) {
+                    updateWorkingRecipe(recipe);
+                    return true;
+                }
+                this.workingRecipe = null;
+                return false;
+            }
+            var items = recipe.getOutputContents(ItemRecipeCapability.CAP);
+            if (!items.isEmpty()) {
+                Map<RecipeCapability<?>, List<Content>> out = Map.of(ItemRecipeCapability.CAP, items);
+                recipe.handleRecipe(io, this.machine, false, out, chanceCaches);
+            }
+            return applyFluidOutputs(recipe, FluidAction.EXECUTE);
+        }
+
+        private boolean applyFluidOutputs(GTRecipe recipe, FluidAction action) {
+            var fluids = recipe.getOutputContents(FluidRecipeCapability.CAP)
+                    .stream()
+                    .map(Content::getContent)
+                    .map(FluidRecipeCapability.CAP::of)
+                    .toList();
+
+            // Distillery recipes should output to the first non-void handler
+            if (recipe.recipeType == GTRecipeTypes.DISTILLERY_RECIPES) {
+                var fluid = fluids.get(0).getStacks()[0];
+                var handler = getMachine().getFirstValid();
+                if (handler == null) return false;
+                int filled = (handler instanceof NotifiableFluidTank nft) ? nft.fillInternal(fluid, action) :
+                        handler.fill(fluid, action);
```

Comment:
Similar small optional formatting suggestion:

```suggestion
                int filled = (handler instanceof NotifiableFluidTank nft) ?
                        nft.fillInternal(fluid, action) :
                        handler.fill(fluid, action);
```